sample/net/dummy/connector/replica-connector.go

```diff
@@ -49,15 +49,53 @@ func (c *ReplicaConnector) ReplicaMessageStreamHandler(replicaID uint32) (api.Me
 	return replica, nil
 }
 
-// ConnectReplica assigns a replica instance to replica ID.
-func (c *ReplicaConnector) ConnectReplica(replicaID uint32, replica api.MessageStreamHandler) {
-	c.replicas[replicaID] = replica
+// ConnectReplica initializes a messageStreamHandler instance.
+func (c *ReplicaConnector) ConnectReplica(replicaID uint32, ready chan bool) {
+	c.replicas[replicaID] = &messageStreamHandler{replicaID, ready, nil}
```

Comment:
> ... `ReplicaConnector.ReplicaMessageStreamHandler` could create and return an instance of `messageStreamHandler` by itself.

This is what I missed in previous versions :astonished:.  In this approach, we can reduce the lines of changes in `core/integration_test.go`.

> We add a map/slice of `chan bool` or `chan struct{}` in `ReplicaConnector` structure.

Yes, relocating channels into `ReplicaConnector` looks fine to me, but I found that I got warnings from data race detector when doing like below:
<pre>
type ReplicaConnector struct {
    replicas map[uint32]api.MessageStreamHandler
    ready    map[uint32]chan bool
}

...

func (c *ReplicaConnector) ConnectReplica(replicaID uint32, replica api.MessageStreamHandler) {
    c.replicas[replicaID] = replica
    c.ready[replicaID] <- true
}

...

func (h *messageStreamHandler) HandleMessageStream(in <-chan []byte) <-chan []byte {
    out := make(chan []byte)

    go func() {
        <-h.connector.ready[h.replicaID]
        replica := h.connector.replicas[h.replicaID]
        for msg := range replica.HandleMessageStream(in) {
            out <- msg
        }
    }()
</pre>
It seems to me that passing `Replica` instance via channel is better than "write-notify-wake-read" pattern, so I'm thinking of converting the type of `ReplicaConnector.replicas` from ` map[uint32]api.MessageStreamHandler` to `map[uint32]chan api.MessageStreamHandler`.

I have a next version of changeset with your approach, so will update PR soon.

Reply:
I think I know why you experienced data races in that code. Normally, a channel can act as a synchronization primitive. However in your case, the data race happened not because of concurrent access to the values of the keys, but due to the concurrent read/write access to the map itself. Since the total number of replicas are known at construction time, we can just allocate a slice of `api.MessageStreamHandler` and never change the slice itself afterwards. As opposed to maps, it is safe to access different indexes of the same slice concurrently.

Reply:
Thank you for elaborating, I confirmed that I can fix it with using slice.